master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+package searcher
+
+import (
+	"math"
+	"sort"
+
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// AsyncHalvingSearch implements a search using the asynchronous successive halving algorithm
+// (ASHA). The experiment will run until the target number of trials have been completed
+// in the bottom rung and no further promotions can be made to higher rungs.
+type asyncHalvingSearch struct {
+	defaultSearchMethod
+	model.AsyncHalvingConfig
+
+	rungs      []*rung
+	trialRungs map[RequestID]int
+	// earlyExitTrials contains trials that exited early that are still considered in the search.
+	earlyExitTrials     map[RequestID]bool
+	totalStepsCompleted int
+	maxTrials           int
+	trialsCompleted     int
+}
+
+const ashaExitedMetricValue = math.MaxFloat64
+
+func newAsyncHalvingSearch(config model.AsyncHalvingConfig) SearchMethod {
+	rungs := make([]*rung, 0, config.NumRungs)
+	for id := 0; id < config.NumRungs; id++ {
+		compound := math.Pow(config.Divisor, float64(config.NumRungs-id-1))
+		stepsNeeded := max(int(float64(config.TargetTrialSteps)/compound), 1)
+		rungs = append(rungs,
+			&rung{
+				stepsNeeded:       stepsNeeded,
+				outstandingTrials: 0})
+	}
+
+	return &asyncHalvingSearch{
+		AsyncHalvingConfig: config,
+		rungs:              rungs,
+		trialRungs:         make(map[RequestID]int),
+		earlyExitTrials:    make(map[RequestID]bool),
+		maxTrials:          config.MaxTrials,
+	}
+}
+
+// promotions handles bookkeeping of validation metrics and returns a RequestID to promote if
+// appropriate.
+func (r *rung) promotionsAsync(requestID RequestID, metric float64, divisor float64) []RequestID {
+	// See if there is a trial to promote. We are increasing the total number of trials seen by 1; the
+	// number of best trials that definitely should have been promoted so far (numPromote) can only
+	// stay the same or increase by 1.
+	oldNumPromote := int(float64(len(r.metrics)) / divisor)
+	numPromote := int(float64(len(r.metrics)+1) / divisor)
+
+	// Insert the new trial result in the appropriate place in the sorted list.
+	insertIndex := sort.Search(
+		len(r.metrics),
+		func(i int) bool { return r.metrics[i].metric > metric },
+	)
+	promoteNow := insertIndex < numPromote
+
+	r.metrics = append(r.metrics, trialMetric{})
+	copy(r.metrics[insertIndex+1:], r.metrics[insertIndex:])
+	r.metrics[insertIndex] = trialMetric{
+		requestID: requestID,
+		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
+	}
+
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
+		return []RequestID{requestID}
+	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
+		if !t.promoted {
+			t.promoted = true
```

Comment:
non-blocking: I think it's a bit safer to set this field once we have actually promoted it (in `promoteAsync`) but open to other perspectives if you think different

Reply:
the main reason is that it's easiest to do here since we need to know the trial metric index on the rung which is available here but would need to be passed outside the promotionsAsync function in order to set promote in promoteAsync.  in terms of safety, i don't think it's that much safer to set it in promoteAsync than in promotionsAsync since promoteAsync shouldn't error.

Reply:
yeah I don't think it's a big deal, my thought was that we can set promoted when we loop through the output. I don't think we'd need to pass the index since we are already passing the object as a return. It's non-blocking, so I will leave it up to you to do what you think is best 👍

Reply:
is there any asynchrony in how searcher processes events?